### ISSUE TITLE
Builds non-deterministically fail with "Couldn't generate classes"

### ISSUE THREAD

**Issue by seanabraham** [evernote/android-state#35]
We have a large, multi-module project (~80ish modules) and the javac annotation processing step occasionally fails when using `android-state` with "error: Couldn't generate classes". There are many other annotation processors that also run without issue. They include Dagger, Epoxy, AutoValue, Butterknife, etc.

I'm assuming the error comes from here but I'm not clear on which line in the block is throwing the exception: https://github.com/evernote/android-state/blob/master/processor/src/main/java/com/evernote/android/state/StateProcessor.java#L467

Any thoughts or ideas for workarounds?

**Comment by vRallev** [comment]
I'm sorry about the issue. I've published a new snapshot build with better log messages. Maybe that helps to identify the problem. Do the following to use the snapshot:
```
repositories {
    jcenter()

    maven {
        url 'https://oss.sonatype.org/content/repositories/snapshots/'
    }
}

dependencies {
    compile 'com.evernote:android-state:1.1.6-SNAPSHOT'
    kapt 'com.evernote:android-state-processor:1.1.6-SNAPSHOT' // or annotationProcessor if not Kotlin
}
```

**Comment by seanabraham** [comment]
Alright, I just kicked off 20 builds on CI and was able to get a failure, this is what the stacktrace looks like:
```

java.util.NoSuchElementException
--
  | at java.util.Scanner.throwFor(Scanner.java:862)
  | at java.util.Scanner.next(Scanner.java:1371)
  | at com.evernote.android.state.StateProcessor.getLicenseHeader(StateProcessor.java:928)
  | at com.evernote.android.state.StateProcessor.writeJavaFile(StateProcessor.java:454)
  | at com.evernote.android.state.StateProcessor.process(StateProcessor.java:440)
  | at com.facebook.buck.jvm.java.TracingProcessorWrapper.process(TracingProcessorWrapper.java:98)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:826)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:742)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2000(JavacProcessingEnvironment.java:100)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1046)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1153)
  | at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1120)
  | at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1018)
  | at com.sun.tools.javac.api.JavacTaskImpl.enter(JavacTaskImpl.java:315)
  | at com.sun.tools.javac.api.JavacTaskImpl.enter(JavacTaskImpl.java:257)
  | at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  | at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  | at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  | at java.lang.reflect.Method.invoke(Method.java:498)
  | at com.facebook.buck.jvm.java.plugin.adapter.BuckJavacTask.enter(BuckJavacTask.java:93)
  | at com.facebook.buck.jvm.java.plugin.adapter.BuckJavacTaskProxyImpl.enter(BuckJavacTaskProxyImpl.java:63)
  | at com.facebook.buck.jvm.java.Jsr199JavacInvocation.invokeCompiler(Jsr199JavacInvocation.java:219)
  | at com.facebook.buck.jvm.java.Jsr199JavacInvocation.buildClasses(Jsr199JavacInvocation.java:177)
  | at com.facebook.buck.jvm.java.JavacStep.execute(JavacStep.java:149)
  | at com.facebook.buck.step.DefaultStepRunner.runStepForBuildTarget(DefaultStepRunner.java:45)
  | at com.facebook.buck.rules.CachingBuildRuleBuilder$BuildRuleSteps.executeCommandsNowThatDepsAreBuilt(CachingBuildRuleBuilder.java:1682)
  | at com.facebook.buck.rules.CachingBuildRuleBuilder$BuildRuleSteps.run(CachingBuildRuleBuilder.java:1638)
  | at com.facebook.buck.rules.BuildRulePipelinesRunner$BuildRulePipelineStage.run(BuildRulePipelinesRunner.java:230)
  | at com.facebook.buck.rules.BuildRulePipelinesRunner$BuildRulePipeline.run(BuildRulePipelinesRunner.java:152)
  | at com.facebook.buck.rules.BuildRulePipelinesRunner$1.run(BuildRulePipelinesRunner.java:103)
  | at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$2(WeightedListeningExecutorService.java:104)
  | at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$0(WeightedListeningExecutorService.java:78)
  | at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:211)
  | at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:200)
  | at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:130)
  | at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:988)
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  | at java.lang.Thread.run(Thread.java:748)

```

And the error is reported as something like `Couldn't generate classes com.airbnb.android.foo.bar.BazFragment$$StateSaver`

I'm going to look a bit more closely at the particular failing fragment to see if anything is amiss (but since this is non-deterministic, I'm not sure what I'd be looking for).

Any thoughts on the stacktrace @vRallev?

**Comment by vRallev** [comment]
This was really helpful. I've published another build. This time your build shouldn't crash anymore, but you should see a warning instead. The annotation processor comes with a license header. For some reason this file cannot be read from the .jar. Please let me know if that works for you.

**Comment by seanabraham** [comment]
@vRallev 
I don't fully understand how all the pieces work but I can hypothesize about what was causing the crash before, depending on the implementation of `getResourceAsStream(String);`. It's a little unclear what exactly that method does, but it looks like it's delegated to the class loader. If there's any case in which subsequent calls to that method return the _same instance_ I believe what may have been happening was the following:

1) Module Foo reaches try-with-resources block, gets instance A of the `inputStream`, opens a Scanner with that stream.
2) Module Bar reaches try-with-resources block, also gets instance A of the `inputStream`, opens another Scanner with that stream
3) Module Foo exits the try-with-resources block, closes the Scanner therefore closing instance A of the `inputStream`
4) Module Bar calls `next()` via the Scanner but since the the `inputStream` is already closed, we get the crash we saw above.

If that's correct, I believe the same might happen even in the case of using a `BufferedReader` instead of `Scanner`. In any case, thank you for making it a warning and allowing the build to succeed! Out of curiosity, I'll use the new release and run it a bunch of times to see if I can get the warning to pop up which would kinda validate the theory (but at least the build won't fail!)

Thanks again for the quick PR!

**Comment by vRallev** [comment]
I'm not sure if that explains it, because creating the stream is inside of a synchronized block. Let me know what you find out.

**Comment by seanabraham** [comment]
@vRallev, I haven't been able to get the warning yet, I'll let you know if I see anything pop up later though. Everything is looking promising so far! In the meantime, do you plan on cutting a release with the fix in it soon? Just want to know if I should cut an internal release before then.

**Comment by vRallev** [comment]
I'll be in London starting tomorrow. I plan to release a new version asap, but not before Monday next week.

**Comment by seanabraham** [comment]
Any word on that release @vRallev?

**Comment by seanabraham** [comment]
Ah, I have another update on this @vRallev!

I've run some builds with the latest code here and found a case that still breaks.

Similar to before, the stream closes but we get a warning this time instead of an error which is nice:
```

java.io.IOException: Stream closed
--
  | at java.util.zip.InflaterInputStream.ensureOpen(InflaterInputStream.java:67)
  | at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:142)
  | at java.io.FilterInputStream.read(FilterInputStream.java:133)
  | at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
  | at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
  | at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
  | at java.io.InputStreamReader.read(InputStreamReader.java:184)
  | at java.io.BufferedReader.fill(BufferedReader.java:161)
  | at java.io.BufferedReader.readLine(BufferedReader.java:324)
  | at java.io.BufferedReader.readLine(BufferedReader.java:389)
  | at com.evernote.android.state.StateProcessor.getLicenseHeader(StateProcessor.java:928)
  | at com.evernote.android.state.StateProcessor.writeJavaFile(StateProcessor.java:455)
  | at com.evernote.android.state.StateProcessor.process(StateProcessor.java:441)
  | at com.facebook.buck.jvm.java.TracingProcessorWrapper.process(TracingProcessorWrapper.java:98)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:826)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:742)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2000(JavacProcessingEnvironment.java:100)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1046)
  | at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1153)
  | at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1120)
  | at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1018)
  | at com.sun.tools.javac.api.JavacTaskImpl.enter(JavacTaskImpl.java:315)
  | at com.sun.tools.javac.api.JavacTaskImpl.enter(JavacTaskImpl.java:257)
  | at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  | at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  | at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  | at java.lang.reflect.Method.invoke(Method.java:498)
  | at com.facebook.buck.jvm.java.plugin.adapter.BuckJavacTask.enter(BuckJavacTask.java:93)
  | at com.facebook.buck.jvm.java.plugin.adapter.BuckJavacTaskProxyImpl.enter(BuckJavacTaskProxyImpl.java:63)
  | at com.facebook.buck.jvm.java.Jsr199JavacInvocation.invokeCompiler(Jsr199JavacInvocation.java:219)
  | at com.facebook.buck.jvm.java.Jsr199JavacInvocation.buildClasses(Jsr199JavacInvocation.java:177)
  | at com.facebook.buck.jvm.java.JavacStep.execute(JavacStep.java:149)
  | at com.facebook.buck.step.DefaultStepRunner.runStepForBuildTarget(DefaultStepRunner.java:45)
  | at com.facebook.buck.rules.CachingBuildRuleBuilder$BuildRuleSteps.executeCommandsNowThatDepsAreBuilt(CachingBuildRuleBuilder.java:1682)
  | at com.facebook.buck.rules.CachingBuildRuleBuilder$BuildRuleSteps.run(CachingBuildRuleBuilder.java:1638)
  | at com.facebook.buck.rules.BuildRulePipelinesRunner$BuildRulePipelineStage.run(BuildRulePipelinesRunner.java:230)
  | at com.facebook.buck.rules.BuildRulePipelinesRunner$BuildRulePipeline.run(BuildRulePipelinesRunner.java:152)
  | at com.facebook.buck.rules.BuildRulePipelinesRunner$1.run(BuildRulePipelinesRunner.java:103)
  | at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$2(WeightedListeningExecutorService.java:104)
  | at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$0(WeightedListeningExecutorService.java:78)
  | at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:211)
  | at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:200)
  | at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:130)
  | at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:988)
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  | at java.lang.Thread.run(Thread.java:748)
```

Leads to:
`warning: Couldn't read the license file`

However, we get another error since somewhere in place of the license, "null" is written to the top of the file, something like:
```
/path/to/file/com/foo/bar/Baz$$StateSaver.java:1: error: class, interface, or enum expected
--
  | nullpackage com.foo.bar;
```

so the build still fails.

**Comment by vRallev** [comment]
I'm sorry for the delay. I've released version `1.1.6`. I've moved the license header now into the code as a String constant. That should fix the problem. Let me know if you run into other issues. Thanks for your help!

**Comment by seanabraham** [comment]
Thank you so much @vRallev!
